display_list/effects/dl_image_filter.h

```diff
@@ -741,6 +750,82 @@ class DlLocalMatrixImageFilter final : public DlImageFilter {
   std::shared_ptr<const DlImageFilter> image_filter_;
 };
 
+class DlRuntimeEffectImageFilter final : public DlImageFilter {
+ public:
+  explicit DlRuntimeEffectImageFilter(
+      sk_sp<DlRuntimeEffect> runtime_effect,
+      std::vector<std::shared_ptr<DlColorSource>> samplers,
+      std::shared_ptr<std::vector<uint8_t>> uniform_data)
+      : runtime_effect_(std::move(runtime_effect)),
+        samplers_(std::move(samplers)),
+        uniform_data_(std::move(uniform_data)) {}
+
+  std::shared_ptr<DlImageFilter> shared() const override {
+    return std::make_shared<DlRuntimeEffectImageFilter>(
+        this->runtime_effect_, this->samplers_, this->uniform_data_);
+  }
+
+  static std::shared_ptr<DlImageFilter> Make(
+      sk_sp<DlRuntimeEffect> runtime_effect,
+      std::vector<std::shared_ptr<DlColorSource>> samplers,
+      std::shared_ptr<std::vector<uint8_t>> uniform_data) {
+    return std::make_shared<DlRuntimeEffectImageFilter>(
+        std::move(runtime_effect), std::move(samplers),
+        std::move(uniform_data));
+  }
+
+  DlImageFilterType type() const override {
+    return DlImageFilterType::kRuntimeEffect;
+  }
+  size_t size() const override { return sizeof(*this); }
+
+  bool modifies_transparent_black() const override { return false; }
+
+  SkRect* map_local_bounds(const SkRect& input_bounds,
+                           SkRect& output_bounds) const override {
+    output_bounds = input_bounds;
+    return &output_bounds;
+  }
+
+  SkIRect* map_device_bounds(const SkIRect& input_bounds,
+                             const SkMatrix& ctm,
+                             SkIRect& output_bounds) const override {
+    output_bounds = input_bounds;
+    return &output_bounds;
+  }
+
+  SkIRect* get_input_device_bounds(const SkIRect& output_bounds,
+                                   const SkMatrix& ctm,
+                                   SkIRect& input_bounds) const override {
+    input_bounds = output_bounds;
+    return &input_bounds;
+  }
+
+  const DlRuntimeEffectImageFilter* asRuntimeEffectFilter() const override {
+    return this;
+  }
+
+  const sk_sp<DlRuntimeEffect> runtime_effect() const {
+    return runtime_effect_;
+  }
+
+  const std::vector<std::shared_ptr<DlColorSource>>& samplers() const {
+    return samplers_;
+  }
+
+  const std::shared_ptr<std::vector<uint8_t>>& uniform_data() const {
+    return uniform_data_;
+  }
+
+ protected:
+  bool equals_(const DlImageFilter& other) const override { return false; }
```

Comment:
Can't they be equal if they have the reference to the same DlRE and the samplers and uniforms all match? Could this affect our determining of what has changed between frames?

Reply:
Updated